### ISSUE TITLE
Login requires captcha inspite of using MFA

### ISSUE THREAD

**Issue by dnlm** [thorsten-gehrig/alexa-remote-control#101]
Amazon for some reason forced me to change my password and I forgot to set the new password in the script's config. When I noticed that it stopped working I set the new password in the config, ran the script again and it said "login failed, check /tmp/.alexa.login". 

This file looks like amazon wants me to enter a captcha, although I'm using MFA. The Password and MFA_Secret set in the config are correct (I logged in with a browser using the generated OTP from oathtool). When I try to log in with a browser (Chromium & Firefox) with JS disabled from the same machine using XMING (it's a raspberry pi zero w/o GUI) no captcha is required.

Logging in via Browser, extracting the cookie and manually saving it to /tmp/.alexa.cookie works but obviously only until the cookie expires. I'd really like to be able to log in automatically again, is there anything else I could try?

**Comment by Apollon77** [comment]
Amazon decides when they want to get presented which security measure. They got very strict with this in the last time. I would have no idea what the script can do against that ...

**Comment by kami83** [comment]
Hi,

i have got the same problem. Maybe someone find a solution. Thanks a lot.

Cu kami

**Comment by matthewbarr** [comment]
I had the same issue.  It's not perfect, but you can make it work with the cookie approach from : https://github.com/thorsten-gehrig/alexa-remote-control/issues/10#issuecomment-399745225

**Comment by kami83** [comment]
Hi, thanks a lot. I have seen this, but you have to renew it on your own every 14 days.

Or?

BG kami

**Comment by dnlm** [comment]
Yes, as cookies have an expiration date. I'm doing the same right now but it's a major inconvenience TBH :(

Does anything (except the effort) prevent a script from extracting the captcha, showing its uri in the terminal for copy/paste purposes and entering the captcha's solution in the terminal again to submit it? Granted it's a simple alphanumerical capture.

I might be able to code this but if anyone already tried it and failed for an obvious reason, I'd like to know.

**Comment by adn77** [comment]
@dnlm what's weird is that you don't get asked for a captcha when you use a browser with JS disabled.
You could try adjusting the user-agent (`BROWSER` env var to match taht of your actual browser).

Maybe creating a new MFA token would somehow reset your "bad-login" count with Amazon as well.

Regarding the captcha - I tried extracting using imageMagick and tesseract. none of the results had been fruitful though :(

**Comment by dnlm** [comment]
> You could try adjusting the user-agent (`BROWSER` env var to match taht of your actual browser).
Gonna try that
 
> Maybe creating a new MFA token would somehow reset your "bad-login" count with Amazon as well.
I already did that but Amazon supplied the same code for generating OTPs so it was useless.

> Regarding the captcha - I tried extracting using imageMagick and tesseract. none of the results had been fruitful though :(
My idea was to extract the captcha image url, print it in the shell and wait for solved captcha user input. Copy, paste in browser and enter captcha solution manually, no image manipulation needed. Or did I misunderstand you?

**Comment by adn77** [comment]
@dnlm I actually tried solving the captcha by using tesseract - and failed miserably ::(

I didn't think about solving the captcha manually. The thing is the captcha is autocreated, you can only download it once. Also, it's a little besides the point of the script.

**Comment by dnlm** [comment]
@adn77 Maybe it was a misconception on my part but I hoped that solving it once on _that_ machine might unlock captchaless logins in the future. For me at least it worked flawlessly until a forced password change which I thought might have triggered a "enter captcha once so we can be sure everything is alright"-mechanic.

Still got to try adjusting the user-agent, will do now and report.

**Comment by dnlm** [comment]
GREAT SUCCESS!!!11 👍 @kami83 @adn77 @matthewbarr 

Changing the user agent sadly didn't work _but_ I tried chromium (js disabled globally) via xming again and got the captcha request. Solved it, logged out, deleted cookies (obviously very important) and logged back in. Again, captcha reappears. Solved it, repeated log out/cookie delete etc and eventually after some tries I wasn't asked for captchas anymore. After logging in 2 times without a captcha request I tried alexa-remote-control on the command line and got logged in immediately. I hope the captcha flag got removed from my account (?) now. If it fails again, I'll report immediately but for now I'll call it solved.

I'm pretty sure it only needed 2 or 3 tries but I mistyped my password on 2 occasions in the process before switching to copy/paste from the password manager 😛 

Also I'm not really sure if using the browser on the same physical machine really is necessary, when I first got the captcha on the pi I tried no-js incognito on my local windows pc and also got the captcha request. But I didn't want to take any chances (regretting it now) and suffered through the painfully slow chromium on my zero w.

I hope this is a permanent solution and might help other people with the same problem.

**Comment by kami83** [comment]
Hi, thanks a lot. Works for me, too.

Cu kami

**Comment by davthomaspilot** [comment]
I'm using oathtool for MFA.  

alexa_remote_control.sh worked for a while, but now fails with the unsuccessful login message.   I do NOT get a captcha when I log in with Chrome.

What do I look for in /tmp/.alexa.login to determine if I'm getting a captcha when I run the alexa_remote_control.sh?  Do I just look at the raw html, or is there a way to figure out what when wrong by using a browser to load that file?

In short, how to I debug what's going wrong?

**Comment by dnlm** [comment]
It failed to login again on its own after the cookie expired. Sadly this means I'll be giving up on this :( really liked the cli approach, bummer

**Comment by dnlm** [comment]
Started working again after about 14 days without any action on my side, will keep reporting if anyone is really interested.

**Comment by dcaccount** [comment]
> Logging in via Browser, extracting the cookie and manually saving it to /tmp/.alexa.cookie works but obviously only until the cookie expires. I'd really like to be able to log in automatically again, is there anything else I could try?

Hello,
can you please advise how to log in via Browser and extract the cookie manually?

Thanks!
Dan

**Comment by dnlm** [comment]
@dcaccount use this extension: https://chrome.google.com/webstore/detail/get-cookiestxt/bgaddhkoddajcdgocldbbfleckgcbcid

And follow this short how-to: https://github.com/thorsten-gehrig/alexa-remote-control/issues/10#issuecomment-381449803 (or the one 4 comments below)

**Comment by dcaccount** [comment]
> @dcaccount use this extension: https://chrome.google.com/webstore/detail/get-cookiestxt/bgaddhkoddajcdgocldbbfleckgcbcid
> 
> And follow this short how-to: [#10 (comment)](https://github.com/thorsten-gehrig/alexa-remote-control/issues/10#issuecomment-381449803) (or the one 4 comments below)

Thanks, I have installed the extension and downloaded but I have a list of cookies in the downloaded txt file.

What one shall I take?

Please note that I deleted previous cookies.

Thanks for helping,
Dan

**Comment by dnlm** [comment]
> What one shall I take? I was looking for a cookie starting with:
> 
> >

IIRC I just pasted the whole file into /tmp/.alexa.cookie and it worked (for a week or two)

**Comment by dcaccount** [comment]
> > What one shall I take? I was looking for a cookie starting with:
> > >
> 
> IIRC I just pasted the whole file into /tmp/.alexa.cookie and it worked (for a week or two)

Thanks!

**Comment by adn77** [comment]
check out the latest feature which doesn't rely on username/password/mfa but uses the refresh_token returned by proper device registration: https://github.com/adn77/alexa-cookie-cli

or more on my blog: https://blog.loetzimmer.de/2021/09/alexa-remote-control-shell-script.html

**Comment by dnlm** [comment]
@adn77 Will try that ASAP, thank you very much for commenting here!

**Comment by adn77** [comment]
Try `echo 'hey you' | grep -E '\sy'` which should print *hey you*

If that's the case, please checkout the latest version, I made a slight change to the "grep" commands.

**Comment by dcaccount** [comment]
> Try `echo 'hey you' | grep -E '\sy'` which should print _hey you_
> 
> If that's the case, please checkout the latest version, I made a slight change to the "grep" commands.

It works, it is awesome!

Thanks a lot.
